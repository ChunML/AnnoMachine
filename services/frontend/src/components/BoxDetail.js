import React from 'react';
import PropTypes from 'prop-types';
import { FaEye, FaCheck, FaEdit, FaTrash } from 'react-icons/fa';
import BoxCoord from './BoxCoord';
import BoxEditForm from './BoxEditForm';
import Label from './Label';
import LabelEditForm from './LabelEditForm';

class BoxDetail extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      boxIsDrawn: props.boxIsDrawn,
      editMode: props.editMode || false,
      coords: {
        id: props.box.id,
        label: props.box.label,
        x_min: Math.floor(props.box.x_min),
        y_min: Math.floor(props.box.y_min),
        x_max: Math.floor(props.box.x_max),
        y_max: Math.floor(props.box.y_max),
      },
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleEyeIconClick = this.handleEyeIconClick.bind(this);
    this.handleCheckIconClick = this.handleCheckIconClick.bind(this);
    this.handleTrashIconClick = this.handleTrashIconClick.bind(this);
    this.toggleEditMode = this.toggleEditMode.bind(this);
  }

  handleInputChange(e) {
    const { name, value } = e.target;
    const { onInputChange } = this.props;
    this.setState(
      prev => ({
        coords: {
          ...prev.coords,
          [name]: name === 'label' ? value : parseFloat(value) || 0,
        },
      }),
      () => {
        const { coords } = this.state;
        onInputChange(coords);
      }
    );
  }

  handleEyeIconClick() {
    const { box, onEyeIconClick } = this.props;
    this.setState(prev => ({
      boxIsDrawn: !prev.boxIsDrawn,
    }));

    onEyeIconClick(box);
  }

  handleCheckIconClick() {
    const { onCheckIconClick } = this.props;
    const { coords } = this.state;
    this.toggleEditMode();
    onCheckIconClick(coords);
  }

  handleTrashIconClick() {
    const { onTrashIconClick } = this.props;
    const { coords } = this.state;
    onTrashIconClick(coords.id);
  }

  toggleEditMode() {
    this.setState(prev => ({
      editMode: !prev.editMode,
    }));
  }

  render() {
    const { box } = this.props;

    const { boxIsDrawn, editMode, coords } = this.state;

    return (
      <div className="segment" key={box.id}>
        <div className="grid">
          <div className="grid-row space-vertical">
            <div className="flex-center space-around">
              <div>
                {editMode ? (
                  <LabelEditForm
                    label={box.label}
                    onInputChange={this.handleInputChange}
                  />
                ) : (
                  <Label label={box.label} />
                )}
              </div>
              <div>
                <span>
                  <button
                    className={`circular ${boxIsDrawn ? 'primary' : ''} button`}
                    onClick={this.handleEyeIconClick}
                  >
                    <FaEye />
                  </button>
                  {editMode ? (
                    <button
                      className="circular primary button"
                      onClick={this.handleCheckIconClick}
                    >
                      <FaCheck />
                    </button>
                  ) : (
                    <button
                      className="circular red button"
                      onClick={this.toggleEditMode}
                    >
                      <FaEdit color="white" />
                    </button>
                  )}
                  <button
                    className="circular red button"
                    onClick={this.handleTrashIconClick}
                  >
                    <FaTrash color="white" />
                  </button>
                </span>
              </div>
            </div>
          </div>
          <div className="grid-row space-vertical">
            {editMode ? (
              <BoxEditForm
                coords={coords}
                onInputChange={this.handleInputChange}
              />
            ) : (
              <BoxCoord coords={coords} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

BoxDetail.propTypes = {
  box: PropTypes.object.isRequired,
  boxIsDrawn: PropTypes.bool.isRequired,
  editMode: PropTypes.bool,
  onInputChange: PropTypes.func.isRequired,
  onEyeIconClick: PropTypes.func.isRequired,
  onCheckIconClick: PropTypes.func.isRequired,
  onTrashIconClick: PropTypes.func.isRequired,
};

BoxDetail.defaultProps = {
  editMode: false,
};

export default BoxDetail;
