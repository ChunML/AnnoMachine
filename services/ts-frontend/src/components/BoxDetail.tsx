import React, { useEffect, useState } from "react";
import { FaEye, FaCheck, FaEdit, FaTrash } from "react-icons/fa";
import BoxCoord from "./BoxCoord";
import BoxEditForm from "./BoxEditForm";
import Label from "./Label";
import LabelEditForm from "./LabelEditForm";
import { BoxType } from "../App";

interface BoxDetailProps {
  boxIsDrawn: boolean;
  editMode: boolean;
  box: BoxType;
  onInputChange(coords: BoxType): void;
  onEyeIconClick(coords: BoxType): void;
  onCheckIconClick(coords: BoxType): void;
  onTrashIconClick(id: number): void;
}

interface BoxDetailStates {
  boxIsDrawn: boolean;
  editMode: boolean;
  coords: BoxType;
}

function BoxDetail(props: BoxDetailProps) {
  const [boxIsDrawn, setBoxIsDrawn] = useState<boolean>(props.boxIsDrawn);
  const [editMode, setEditMode] = useState<boolean>(props.editMode || false);
  const [coords, setCoords] = useState<BoxType>({
    id: props.box.id,
    label: props.box.label,
    x_min: Math.floor(props.box.x_min),
    y_min: Math.floor(props.box.y_min),
    x_max: Math.floor(props.box.x_max),
    y_max: Math.floor(props.box.y_max),
  });

  useEffect(() => {
    setCoords({
      id: props.box.id,
      label: props.box.label,
      x_min: Math.floor(props.box.x_min),
      y_min: Math.floor(props.box.y_min),
      x_max: Math.floor(props.box.x_max),
      y_max: Math.floor(props.box.y_max),
    });
  }, [props.box]);

  const handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { name, value } = e.currentTarget;
    const { onInputChange } = props;
    const newCoords = {
      ...coords,
      [name]: name === "label" ? value : parseFloat(value) || 0,
    };
    onInputChange(newCoords);
  };

  const handleEyeIconClick = () => {
    const { box, onEyeIconClick } = props;
    setBoxIsDrawn(!boxIsDrawn);

    onEyeIconClick(box);
  };

  const handleCheckIconClick = () => {
    const { onCheckIconClick } = props;
    toggleEditMode();
    onCheckIconClick(coords);
  };

  const handleTrashIconClick = () => {
    const { onTrashIconClick } = props;
    onTrashIconClick(coords.id);
  };

  const toggleEditMode = () => {
    setEditMode(!editMode);
  };

  const { box } = props;

  return (
    <div className="segment" key={box.id}>
      <div className="grid">
        <div className="grid-row space-vertical">
          <div className="flex-center space-around">
            <div>
              {editMode ? (
                <LabelEditForm
                  label={box.label}
                  onInputChange={handleInputChange}
                />
              ) : (
                <Label label={box.label} />
              )}
            </div>
            <div>
              <span>
                <button
                  className={`circular ${boxIsDrawn ? "primary" : ""} button`}
                  onClick={handleEyeIconClick}
                >
                  <FaEye />
                </button>
                {editMode ? (
                  <button
                    className="circular primary button"
                    onClick={handleCheckIconClick}
                  >
                    <FaCheck />
                  </button>
                ) : (
                  <button
                    className="circular red button"
                    onClick={toggleEditMode}
                  >
                    <FaEdit color="white" />
                  </button>
                )}
                <button
                  className="circular red button"
                  onClick={handleTrashIconClick}
                >
                  <FaTrash color="white" />
                </button>
              </span>
            </div>
          </div>
        </div>
        <div className="grid-row space-vertical">
          {editMode ? (
            <BoxEditForm coords={coords} onInputChange={handleInputChange} />
          ) : (
            <BoxCoord coords={coords} />
          )}
        </div>
      </div>
    </div>
  );
}

export default BoxDetail;
