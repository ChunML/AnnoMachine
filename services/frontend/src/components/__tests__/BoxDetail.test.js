import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import BoxDetail from '../BoxDetail';
import '../../setupTests';

const box = {
  id: 1,
  label: 'dog',
  x_min: 10,
  y_min: 10,
  x_max: 20,
  y_max: 20,
};

describe('Not in edit mode', () => {
  const component = (
    <BoxDetail
      box={box}
      boxIsDrawn={false}
      onInputChange={jest.fn()}
      onEyeIconClick={jest.fn()}
      onCheckIconClick={jest.fn()}
      onTrashIconClick={jest.fn()}
    />
  );

  it('BoxDetail renders properly', () => {
    const wrapper = shallow(component);
    const grid = wrapper.find('.grid');
    expect(grid.length).toBe(1);
    const rows = grid.find('.grid-row');
    expect(rows.length).toBe(2);
    const columns = rows.at(0).find('.flex-center > div');
    expect(columns.length).toBe(2);
    const buttons = columns.find('button');
    expect(buttons.length).toBe(3);
    expect(buttons.at(0).find('FaEye').length).toBe(1);
    expect(buttons.at(1).find('FaEdit').length).toBe(1);
    expect(buttons.at(2).find('FaTrash').length).toBe(1);
    expect(component.props.onEyeIconClick).toHaveBeenCalledTimes(0);
    expect(component.props.onTrashIconClick).toHaveBeenCalledTimes(0);
    buttons.at(0).simulate('click');
    expect(component.props.onEyeIconClick).toHaveBeenCalledTimes(1);
    buttons.at(2).simulate('click');
    expect(component.props.onTrashIconClick).toHaveBeenCalledTimes(1);
  });

  it('BoxDetail renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('In edit mode', () => {
  const component = (
    <BoxDetail
      editMode
      box={box}
      boxIsDrawn={false}
      onInputChange={jest.fn()}
      onEyeIconClick={jest.fn()}
      onCheckIconClick={jest.fn()}
      onTrashIconClick={jest.fn()}
    />
  );

  it('BoxDetail renders properly', () => {
    const wrapper = shallow(component);
    const instance = wrapper.instance();
    const grid = wrapper.find('.grid');
    expect(grid.length).toBe(1);
    const rows = grid.find('.grid-row');
    expect(rows.length).toBe(2);
    const columns = rows.at(0).find('.flex-center > div');
    expect(columns.length).toBe(2);
    const buttons = columns.find('button');
    expect(buttons.length).toBe(3);
    expect(buttons.at(0).find('FaEye').length).toBe(1);
    expect(buttons.at(1).find('FaCheck').length).toBe(1);
    expect(buttons.at(2).find('FaTrash').length).toBe(1);
    expect(component.props.onEyeIconClick).toHaveBeenCalledTimes(0);
    expect(component.props.onCheckIconClick).toHaveBeenCalledTimes(0);
    expect(component.props.onTrashIconClick).toHaveBeenCalledTimes(0);
    instance.handleInputChange({
      target: {
        name: 'x_min',
        value: 20,
      },
    });
    instance.handleInputChange({
      target: {
        name: 'label',
        value: 'cat',
      },
    });
    expect(wrapper.state()).toEqual({
      boxIsDrawn: false,
      editMode: true,
      coords: { ...box, x_min: 20, label: 'cat' },
    });
    buttons.at(0).simulate('click');
    expect(component.props.onEyeIconClick).toHaveBeenCalledTimes(1);
    buttons.at(1).simulate('click');
    expect(component.props.onCheckIconClick).toHaveBeenCalledTimes(1);
    buttons.at(2).simulate('click');
    expect(component.props.onTrashIconClick).toHaveBeenCalledTimes(1);
  });

  it('BoxDetail renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
