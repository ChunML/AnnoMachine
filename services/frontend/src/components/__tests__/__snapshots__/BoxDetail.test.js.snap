// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxDetail renders a snapshot properly 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui center aligned two column divided grid"
  >
    <div
      className="row"
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <div
        className="column"
      >
        <span
          className="ui header"
          style={
            Object {
              "color": "firebrick",
              "letterSpacing": "0.25rem",
            }
          }
        >
          dog
        </span>
      </div>
      <div
        className="column"
      >
        <span>
          <button
            className="circular grey ui icon button"
            onClick={[Function]}
          >
            <i
              className="eye icon"
            />
          </button>
          <button
            className="circular red ui icon button"
            onClick={[Function]}
          >
            <i
              className="edit icon"
            />
          </button>
        </span>
      </div>
    </div>
    <div
      className="row"
      style={
        Object {
          "paddingTop": "0",
        }
      }
    >
      <span
        style={
          Object {
            "background": "powderblue",
            "borderRadius": "4px",
            "color": "maroon",
            "letterSpacing": "0.2rem",
            "margin": "5px",
            "padding": "5px 10px",
          }
        }
      >
        10
      </span>
      <span
        style={
          Object {
            "background": "powderblue",
            "borderRadius": "4px",
            "color": "maroon",
            "letterSpacing": "0.2rem",
            "margin": "5px",
            "padding": "5px 10px",
          }
        }
      >
        10
      </span>
      <span
        style={
          Object {
            "background": "powderblue",
            "borderRadius": "4px",
            "color": "maroon",
            "letterSpacing": "0.2rem",
            "margin": "5px",
            "padding": "5px 10px",
          }
        }
      >
        20
      </span>
      <span
        style={
          Object {
            "background": "powderblue",
            "borderRadius": "4px",
            "color": "maroon",
            "letterSpacing": "0.2rem",
            "margin": "5px",
            "padding": "5px 10px",
          }
        }
      >
        20
      </span>
    </div>
  </div>
</div>
`;
